src/verifast1.ml

```diff
@@ -2709,45 +2709,51 @@ module VerifyProgram1(VerifyProgramArgs: VERIFY_PROGRAM_ARGS) = struct
       (i, (l, fieldmap, meths, preds, supers, pn, ilist, tparams))
     end
   
-  let rec lookup_class_field cn fn =
+  let rec lookup_class_field (cn, targs) fn =
+    let replace_type_param {fl; fgh; ft; fvis; fbinding; ffinal; finit; fvalue} tparams = 

```

Comment:
Beter: `replace_type_params`?

Reply:
Ook: je roept deze functie op in de gevallen waarbij het veld gevonden wordt in `cn`, maar niet wanneer het veld gevonden wordt in een supertype van `cn`. In die gevallen is dat toch ook nodig? In plaats van de functie telkens in de specifieke gevallen op te roepen, lijkt het mij beter ze één keer op te roepen op het resultaat van de `match` op regel 2719. (Ken dat resultaat toe aan een variabele.)

Reply:
Ik denk dat dit onoverzichtelijker gaat zijn omdat de recursieve aanroepen al de correcte vervangingen gedaan hebben. Dan krijgen we ofwel een lokaal veld dat nog vervangen moet worden, of een veld dat al vervangen is. Hier gaan we dan nog is replace type params op toepassen?

Reply:
Beschouw:
```java
class Foo<T> {
    T x;
}
class Bar<U> extends Foo<ArrayList<U>> {}
```
Stel nu dat we veld `x` van `Bar<Integer>` willen opzoeken. De recursieve oproep zal type `ArrayList<U>` teruggeven, niet `ArrayList<Integer>`.

Reply:
In dit geval denk ik dat het beter is om de vervanging te doen voordat we de recursieve aanroep doen? Anders gaat dit mis als Bar<T> en Foo<T> zijn.

Als ik me niet vergis krijgen we dan lookup_class_field (Foo, ArrayList<T>) en dit geeft dan een oneindige loop want T = ArrayList<T>? De huidige oplossing is dus om de vervanging eerst te doen zodat we lookup_class_field (Foo, ArrayList<Integer>) zouden krijgen.

Ik denk dat het misschien een goed idee zou zijn om in een toekomstige pull request deze functie te updaten zodat er een correcte locatie binnen komt. In principe zou het niet mis mogen gaan denk ik, maar momenteel is de enige plaats dat we een fout kunnen geven in de class zelf.

Reply:
Ik heb de comment geupdate, als het niet duidelijk is kunnen we het misschien rap bespreken via skype?

Reply:
Volgens mij geeft `T` instantiëren met `ArrayList<T>` geen oneindige lus. Beide aanpakken (instantiëren voor of na de recursieve oproep) werken volgens mij, maar de code die ik gereviewed heb deed volgens mij geen van beide.